### PR TITLE
Make sure DirectoryPathTree does not allow reading outside the root dir

### PR DESCRIPTION
**Pull request by aloubyansky**
An alternative to https://github.com/quarkusio/quarkus/pull/23672

### CODE REVIEW

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java
Comment (Sanne) [inline independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:25]: Are we sure this is comprehensive ? UNC formats are also considered valid "absolute paths" on Windows; some people might even argue that URL is the golden standard.

Comment (Sanne) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:25]: (I'm fine either way - just wondering if this was your intent. Perhaps a doc clarification of what's intended is in order)

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:25]: That would be nice if we wanted to support `PathTree` as a general purpose API, which isn't the case currently. For now it's limited to loading resources on the classpath following the Java classpath resource name format. I'd say the main goal currently is to prevent reading outside the directory tree and this seems to be enough. Actually, this check for absolute path is already beyond that and could probably be removed.

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java
Comment (yrodiere) [inline independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:84]: I would not do this.

This behavior is actually the current one, and something we need to avoid in order to fix #23574. The user was setting a configuration property to something like `/home/yrodiere/workspace/myapp/src/main/resources/myfile.txt`. We used `DirectoryPathTree` to check that the resource exist and is a file (not a directory), and the check passed because `DirectoryPathTree` was accepting absolute paths and interpreting them as relative to the FS root. However, later on when we tried to retrieve the resource from the classpath, it didn't work anymore, because `/home/yrodiere/workspace/myapp/src/main/resources/myfile.txt` obviously isn't a valid path to a resource in the classpath.

Also, if you think about it, this behavior allows for Quarkus configurations that may or may not work depending on where on the filesystem the application code is checked out... which seems very bad. For example, if I set some Quarkus config property to `/home/yrodiere/workspace/myapp/src/main/resources/myfile.txt`, and the string eventually ends up in this method, it will be relativized automatically if I run this code on my machine, but not on yours where the application is located in `/home/aloubyansky/workspace/myapp`.

What you are doing is accepting absolute paths as long as they target a file within the directory tree.
What we should be doing, IMO, is what I did in #23672: whenever we get an absolute path, we interpret it as relative to `dir`, i.e. relative to the root of the classpath. In short absolute paths would be absolute paths *in the classpath*, not in the filesystem. Which seems to be what `DirectoryPathTree` is about anyway?

So the code would become:

```suggestion
            if (absolutePath.startsWith(dir)) {
                return absolutePath.relativize(absolutePath.getRoot()).toString();
            }
```

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java
Comment (yrodiere) [inline independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:93]: I really don't think we should do this... In my opinion, the string being passed as input should be interpreted as a paths to a resource in the classpath, i.e. something that we'd normally pass to `Class#getResource`.

The goal of `DirectoryPathTree` as I understand it is to convert such "path in the classpath" to an actual path on the filesystem.

With that in mind, it doesn't make sense to allow input strings to include the path to `src/main/resources` on the filesystem (like you did further up, see my other comment), or to interpret things such as `..` which are a filesystem concept (and not something that is usually supported for paths to resources in the classpath).

**Review by yrodiere — CHANGES_REQUESTED** [review]
Thanks for working on this @aloubyansky, really appreciated!

While I understand this might be better performance-wise than my PR (#23672), there are a few key differences with my PR that I don't think are going in the right direction: IMO we should interpret input strings as paths to resources in the classpath, and never as paths in the filesystem. See my comments below; maybe they explain this more clearly.

Also... unfortunately, regardless of my opinions, this does not fix #23574, which was the point of my PR.

**Comment by aloubyansky** [comment]
Perhaps resource names should be validated in the QuarkusClassLoader and not get so far down the call tree.

**Comment by aloubyansky** [comment]
That way we would avoid this extra sanitization when loading classes.

**Comment by aloubyansky** [comment]
Ah, you are probably using ApplicationArchive? I'll need to check that. Which should follow the same rules for resource names.

**Comment by yrodiere** [comment]
We're using `DirectoryPathTree` in build steps; I don't know if `QuarkusClassLoader` is available at that point? At least, not with access to application resources? I'm not overly familiar with that, though.

Regardless, I don't really see what we're doing as sanitization, but rather as translation. In `DirectoryPathTree`, we expect resource paths as input, and we need to translate them to filesystem paths. We just need to interpret them correctly, and avoid anything clearly filesystem-related (e.g. interpreting input as a path on the filesystem, interpreting `..` in input, ...).

Even if we validated paths using `QuarkusClassLoader` beforehand, we'd still have to interpret the paths correctly afterwards (and take special care with absolute paths, in particular). Pre-validation might help with `..`, though.

**Comment by yrodiere** [comment]
> Ah, you are probably using ApplicationArchive? I'll need to check that. Which should follow the same rules for resource names.

Yes. This is the relevant piece of code: https://github.com/quarkusio/quarkus/blob/4f81ab8d56142a547100bdc08b91071fd57d9cca/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L1112-L1116

**Comment by aloubyansky** [comment]
Thanks @yrodiere for all the analysis. The PathTree is exposed in too many places now, so we do need to be careful about how the relative paths are resolved in it.

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java
Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:84]: Wouldn't it better to simply throw an `IllegalArgumentException` in case `dir` is not equal to `absolutePath.getRoot()`?
I find it counter intuitive otherwise.

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:84]: Actually, it wouldn't help in your case.

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:84]: It's just there could be two different valid paths: one relative to the actual root of the passed in path and one relative to the `dir`. In which case, it's kind of confusing, especially on Windows, where the absolute path includes the disk part.
Perhaps, we should throw an IllegalArgumentException if we detect the disk part?

Comment (yrodiere) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:84]: I think in my case (#23574), passing an path such as `/my-resource.sql` could make sense: the user might just mean "absolute path to a resource", so "absolute" as in "relative to the root of the classpath". Of course in #23574 that's not what they did; and there's no way to differentiate between a user mistakenly passing an absolute (unix) filesystem path and a user passing an absolute resource path. So I'd say absolute unix paths should probably be allowed and handled like in my suggestion?

Passing a windows absolute path with a disk part, on the other hand... yeah, that doesn't make sense and would deserve an exception.

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java
Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:93]: About this one, I think I'd keep it. The thing is that the PathTree is also exposed as an actual file system path tree API in other places. However, we still should not allow reading outside the tree root.

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:93]: I do agree it's ugly though.

Comment (yrodiere) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:93]: Supporting `..` does raise a problem, though... But maybe one that was pre-existing.

If I pass `foo/bar/..` to `DirectoryPathTree`, it will automatically be resolved to `foo/`. So in a build step for example, we might end up accepting the path as valid.
Later at runtime, we might pass the same string to `Class#getRessource`. And then it will fail, because `..` has no special meaning for `Class#getResource`...

Maybe a better behavior would be to simple escape `..`, so that it's interpreted as a literal folder named `..`? And if that's not possible, maybe we should just reject any path with a component named `..`?

Damn, this is starting to look dodgy. I sure hope there aren't many similar pitfalls with Windows paths.

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:93]: Yes, there could be path element name `a..b`

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java
Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:92]: We should still disallow reading outside the tree.

**Comment by Sanne** [comment]
I wonder if it's wise to *generally* try to load a resource both as a _resource_ and as possibly and absolute path.

Thinking about the use case expressed in #23574 - could we patch that configuration element to explicitly try the `Path` API and then try loading as resource should the first one fail?

**Comment by yrodiere** [comment]
> I wonder if it's wise to _generally_ try to load a resource both as a _resource_ and as possibly and absolute path.
> 
> Thinking about the use case expressed in #23574 - could we patch that configuration element to explicitly try the `Path` API and then try loading as resource should the first one fail?

Personally, I'd object to retrieving files from the filesystem at build time. Especially using absolute paths. It makes the build environment-specific, and I really don't want to encourage that kind of bad practice.

Something relative to the Maven project root... maybe. But then the question becomes how do we make it clear what the user means (resource path or FS path)? Is a try-and-default behavior really a good idea? And also, does that mean we will automatically add a file to the resources, even though it's not in `src/main/resources`?

Really, the more I think about it, the less I like the idea. If people want to use a file that is not a Java resource, in a place where we expect a Java resource, they can tune their build to generate resources (copying stuff to `target/generated/...`). Or *maybe* we can add options to Quarkus to make that easy. But it really should be explicit, IMO.

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java
Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:92]: Yes, that's what I meant. Perhaps after stripping the first `/` I could check again if it still starts with `/` and then throw an IAE.  I suppose it wouldn't be a legal resources name.

**Comment by Sanne** [comment]
> > I wonder if it's wise to _generally_ try to load a resource both as a _resource_ and as possibly and absolute path.
> > Thinking about the use case expressed in #23574 - could we patch that configuration element to explicitly try the `Path` API and then try loading as resource should the first one fail?
> 
> Personally, I'd object to retrieving files from the filesystem at build time. Especially using absolute paths. It makes the build environment-specific, and I really don't want to encourage that kind of bad practice.

I'm not sure I follow. If it's *retrieved* at build time, how is that making the build "environment-specific" ?

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java
Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:92]: It doesn't appear that a resource name can start with `/`. I just tried
````
		Enumeration<URL> resources = Thread.currentThread().getContextClassLoader().getResources("/META-INF");
````
in a non-Quarkus test and it returns an empty enumeration, while it returns 177 URLs for `"META-INF"`. So it looks like we could simply disallow absolute paths.

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java:92]: It could potentially break existing callers/users though.

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: > Perhaps after stripping the first `/` I could check again if it still starts with `/` and then throw an IAE. I suppose it wouldn't be a legal resources name.

I pushed this adjustment. Although I'm still wondering whether we should simply disallow absolute paths completely. Perhaps, simply returning `null`, indicating "it's not visible from within the tree".

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: Yes, I think we should handle absolute paths as not visible. Do you agree @yrodiere ?

**Comment by yrodiere** [comment]
> I'm not sure I follow. If it's _retrieved_ at build time, how is that making the build "environment-specific" ?

Let's say I set a property to `/home/yrodiere/myfile.txt` in my Quarkus `application.properties`. The build will work on my machine (my environment), because `/home/yrodiere/myfile.txt` exists. It won't work in yours, because `/home/yrodiere` doesn't exist. The build is environment-specific: just checking out the code of the application from the git repository is not enough to successfully build the application.

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java
Comment (yrodiere) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: > It doesn't appear that a resource name can start with `/`.

It depends. It can if you use `java.lang.Class#getResource` to retrieve the resource, because by default the path is interpreted as relative to the class. It cannot start with `/` if you use `java.lang.ClassLoader#getResource`, because then the resource name is absolute by default.

But I think it's safe to assume we'll always use `ClassLoader#getResource`; I would personally consider it's a bug if any extension/library uses `Class#getResource` instead.

> Yes, I think we should handle absolute paths as not visible. Do you agree @yrodiere ?

Works for me. As long as users have an actionable error message, it should be fine.

Comment (yrodiere) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: Ah, you meant to return `null`. I'm not sure that's great, because it will leave users wondering why `/META-INF/MANIFEST.MF` tells them the file doesn't exist, while it works when they do `MyClass.getResource("/META-INF/MANIFEST.MF")`, then it works. I know it's not our fault, it's a quirk of Java APIs, but still, it would be nice to avoid that kind of confusion.

Don't you think it would be fine to throw an exception here, because we consider a resource path starting with `/` is invalid? Something like `new ConfigurationError( "Resource paths must not start with '/'" )`. Then users know what's wrong and can solve it very easily.

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: Then I guess we could leave it as it is.

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: My idea was to make consistent with ClassLoader.getResource(name). Tbh, I haven't seen people using Class.getResource(name). It's looking strange to me they don't have the same resource path rules.

Comment (yrodiere) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: > Then I guess we could leave it as it is.

Well the current code will let absolute paths slip through and resolve files outside of the tree.

But you could indeed use something very close to your code. Instead of failing on Windows absolute paths, just fail on all absolute paths, and don't try to handle them at all. Simple and easy.

**Comment by Sanne** [comment]
> The build will work on my machine (my environment), because `/home/yrodiere/myfile.txt` exists. It won't work on yours, because

Hi @yrodiere thanks for explaining and of course I'd agree with that - but it's not what I would expect Quarkus to do.

If the resource if "fixed" at build time, I would expect it to either copy the resource from the filesystem into the resources directory, or read its content and encode it as a constant.

On top of addressing your concern of "environment-specific" it also addresses the semantic aspect of this being in fact a build-time defined constant: if other build-time elements are dependant on its content (e.g. skip some processing because the script is empty), we'd prefer it to be consistent even if you were to edit or remove the file on your local filesystem after the build.

**Comment by yrodiere** [comment]
> If the resource is "fixed" at build time, I would expect it to either copy the resource from the filesystem into the resources directory, or read its content and encode it as a constant (probably depending on the expected reasonable size - I guess the scripts case could be rather large, making it better fit to be treated as resource).
> 
> On top of addressing your concern of "environment-specific" it also addresses the semantic aspect of this being in fact a build-time defined constant: if other build-time elements are dependant on its content (e.g. skip some processing because the script is empty), we'd prefer it to be consistent even if you were to edit or remove the file on your local filesystem after the build.

Yes, I mentioned that solution in a previous message. But this only addresses potential inconsistencies of the environment between build-time and static init (e.g. the file might not be present anymore during static init).

Even if Quarkus copies the file at build time somehow, using an absolute path to a file in the filesystem still looks like bad practice to me: the build is still environment-specific because it relies on some file outside of the scope of your Maven/Gradle project.

And even if we limit this feature to paths relative to the root of your Maven/Gradle project, there are other problems. I'll quote myself:

> Something relative to the Maven project root... maybe. But then the question becomes how do we make it clear what the user means (resource path or FS path)? Is a try-and-default behavior really a good idea? And also, does that mean we will automatically add a file to the resources, even though it was not initially in `src/main/resources`? Because the runtime (e.g. Hibernate ORM) may still expect to retrieve the file with `Class#getResource`...
> 
> Really, the more I think about it, the less I like the idea. If people want to use a file that is not a Java resource, in a place where we expect a Java resource, they can tune their build to generate resources (copying stuff to `target/generated/...`). Or _maybe_ we can add options to Quarkus to make _that_ easy (the copying of stuff to `target/generated` during the build). But it really should be explicit, IMO.

If we still disagree on this, maybe we can have a quick video call later. I'm repeating myself, which is a good indication I'm either failing to understand something or failing to explain something... Either way, sorry about that :sweat_smile:

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java
Comment (yrodiere) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: > Did you see an adjustment I pushed yesterday night? Could you give an example of an absolute path that would slip through?

Ah sorry, I was looking at [ArchivePathTree.java](https://github.com/quarkusio/quarkus/pull/23692/files#diff-0a3f8c0e980f0271b0467b82cba352b591567c97d92ca6b82b6ade7c7d0aeff8), which you didn't adjust, apparently.

So the current code is this:

```java
        if (path.charAt(0) == '/') {
            if (path.length() == 1) {
                return "";
            }
            if (path.charAt(1) == '/') {
                throw new IllegalArgumentException(
                        path + " does not appear to be a path relative to the root of the path tree " + dir);
            }
            return path.substring(1);
        }
```

This solution would have been secure: it allows `/foo/bar` and interprets it as `foo/bar`. It rejects `//foo/bar` and `C:\\foo\bar` with an exception.

However... as you very rightly pointed out, `ClassLoader#getResource` doesn't accept leading slashes. If we do here, we run the risk of accepting resource paths such as `/foo/bar` in build steps, then failing to retrieve the same resource during static init (when calling `ClassLoader#getResource` with the same resource path).

So... it's probably better to just reject all absolute paths?

> I like the idea of not supporting absolute paths. In other cases though, if a path isn't on the tree I let the caller decide whether to throw an exception or ignore. Which is why I'd probably prefer to do the same for absolute paths.

Returning `null` when the user sets an absolute path doesn't seem great to me, because then the caller cannot differentiate between a path that points to a file that doesn't exist (i.e. `foo/doesnotexist.txt`) and a path that is simply incorrect but would point to an existing file if it was relative (i.e. `/foo/exists.txt`). This leads to scenarios where the user sets `/foo/exists.txt`, and gets an error message telling him the file doesn't exist. But it does exist, it's just that we don't understand the path, which is a very different situation!

So, same conclusion: it's probably better to just reject all absolute paths? With a clear exception message telling users that we don't accept absolute paths and they should just make their path relative to their resources (e.g. the `src/main/resources` directory).

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: Ok @yrodiere, done

**Comment by Sanne** [comment]
@yrodiere you have valid points from a perspective of "good engineering" of a stable project, which would dictate scripts need to be part of the project, and versioned together with the rest of the project code.

But I do wonder if you're focusing only on the "production" mindset; we should also be pragmatic for the use case of dev-mode, in which people are exploring and experimenting and it would be at least *useful* to be able to directly point to some scripts which are possibly part of another project and/or being edited (or output) live with different tools that might be open in different windows. In such a scenario, we don't want to get in the way of "the flow" - if possible.

I don't presume knowing what workflow people *should* use, but allowing options _in dev-mode_ shouldn't be considered an anti-pattern to avoid in name of principles.

On the other hand, I'm onboad with such restrictions for a normal (or even a native) build, for which I mentioned the content needs to get "made constant" during a regular build.

**Comment by yrodiere** [comment]
My goal here is to fix a bug (see https://github.com/quarkusio/quarkus/issues/23574#issuecomment-1038972226) and thus I am indeed focusing on the production mindset.

People who want to do weird things and experiment are welcome to open tickets and expose their use case, and their feature request will be addressed when someone agrees with their use case and has time to implement the feature. That's pretty much what I answered [here](https://github.com/quarkusio/quarkus/issues/23574#issuecomment-1038964801).

I the meantime... let's focus on the (already complex) task of having Quarkus produce actionable error messages at build time when it encounters currently unsupported configuration, rather than ignoring it and failing at runtime because of that :)

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java
Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: The issue is that, this API isn't strictly limited to classpath resources, which is why I am hesitating.

**Review by yrodiere — COMMENTED** [review]
@aloubyansky Almost there! LGTM, except the bit about `..`. See https://github.com/quarkusio/quarkus/pull/23692#discussion_r811276732
Once that's addressed, I think we can merge?

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java
Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: I think especially if you are using `ApplicationArchive.getChildPath(path)`, I don't think it's agreed upon that the `path` argument is actually following classloader resource naming format. I think it used be an actual relative FS path before `PathTree` was introduced.
The `PathTree` is also used to represent artifact content of application dependencies.

Comment (yrodiere) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: Hu. I didn't know that.

So we have a single API for multiple types of resource trees, some of which don't have the same syntax for their path strings? That looks like a problem in itself...

Ok, I'll trust you to pick what seems best to you, but to me those are the only possible solutions in the scope of this PR:

* Your implementation, which allows using `..`, which could lead to incorrectly accepting invalid paths for classpath resource trees (see my example above).
* My implementation, which prevents using `..`, which could lead to incorrectly rejecting valid path for non-classpath resource trees.

Alternatively, we could provide a flag in the constructor, to switch between the two behaviors.

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: Yes, before `PathTree` we would simple do `path.resolve(childPath)` in `ApplicationArchive.getChildPath(childPath)`. The primary goal for `PathTree` wasn't actually to represent the classpath resources but to provide a single API to process directory and archive content with the option of applying filters. I would probably keep it that way.
However, I get your point. We actually do have `PathTreeClassPathElement` which implements `ClassPathElement` and its purpose is pretty much what you are asking for. But it's not yet exposed from `ApplicationArchive`. I think it should be, it would have to be a separate refactoring task.
You can actually use `QuarkusClassLoader.getElements(resourceName, localOnly)` which will return a list of `PathTreeClassPathElement`s. But it was added kind of quickly and isn't a nice/proper API for classpath resource locating. Also, it's different from the `ApplicationArchive`, in a sense that `ApplicationArchive`s are a subset of all the dependencies that are identified as such.

**Review by yrodiere — APPROVED** [review]

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java
Comment (yrodiere) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: Alright, I approved the PR; I'm not sure what you're going to do with this `..` handling, so I'll let you merge once you're done?

Thanks again for all the work.

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: I am thinking I probably wouldn't mind the change you proposed. It wouldn't be a huge limitation at this point and we could reconsider supporting `..` in case we really need to one day. Would you like to add this change to your original PR? Thank you @yrodiere

Comment (yrodiere) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java]: Will do. Thanks.

**Pull request merged**
